MoltenVK/MoltenVK/Utility/MVKVector.h

```diff
@@ -524,5 +524,10 @@ template<class Type, class Allocator = mvk_vector_allocator_with_stack<Type, 8>>
   }
 };
 
+
+template<typename Type>
+using MVKVector128 = MVKVector<Type, mvk_vector_allocator_with_stack<Type, 128>>;

```

Comment:
I apologize for this confusion and for not properly explaining what my changes do or intend to do. As for the merge conflicts, sorry also, we do not work with Git, so I am probably not using it properly. I will add my changes again and hopefully properly and try to apply a new pull request after Christmas in case you are interested in merging my changes.

But some more explanations:

As for MVKVector128: Again it's basically a std::vector with the exception, that it has 128 elements in its own class, which it can use, before a heap allocation takes place. That's all. If you need more than 128 elements, MVKVector will reserve this space from the heap, so it can of course grow. It's not a maximum number of elements. Really if you want another name for this, I have no issues with that.

The reason I was talking about 'stack' is when you use this MVKVector in a function:

`void function()
{
  MVKVector<int> test;
  test.push_back( 13 );
}`

In this case, no heap is used, instead the pre allocated elements of MVKVector ( which I defaulted to 8 ) are taken from the stack. So you are right, the pre-allocation takes place in the MVKVector class itself.

The main motivation for using a class like this in MoltenVK is, that in some places std::vector was used dynamically and therefor you do have heap allocations per frame. Again my measurements showed thousands of allocations per frame in our use case. My class will be obsolete in cases where you use the object pool mechanism you mentioned in a previous posting. From what I can see since my initial pull request, you already did make changes to better reuse commands that have std::vector data members. So in this case you could revert back to std::vector.

But there is another reason for using this class:

In case only the pre allocated number of elements are used MVKVector might give you a better cache usage, as accessing the elements avoids a cache miss in some situations.

Facebook's folly for example has a similar vector. If you read the first few sentences you might get a better idea:

https://github.com/facebook/folly/blob/master/folly/docs/small_vector.md

Reply:
Don't worry...it's not confusing. I think it the overall idea is great...and quite useful. Thanks for the background explanation...and for the reference to Folly's `small_vector`.

Does `MVKVector` also meet the requirement of contiguous memory after a spill-over?

I'm fine with all of this...except for the use of dedicated class names...per my explanation above.

Is it possible to move the dedicated size to a constructor parameter...so the size only appears on the line that creates the vector...instead of everywhere the vector is passed?

Reply:
> Don't worry...it's not confusing. I think it the overall idea is great...and quite useful. Thanks for the background explanation...and for the reference to Folly's `small_vector`.

See also [`llvm::SmallVector`](/llvm-mirror/llvm/blob/69ba13fd38262b31264a9447aa2d7d7af95daabd/include/llvm/ADT/SmallVector.h#L847). I mentioned that on the other PR #376.

> 
> I'm fine with all of this...except for the use of dedicated class names...per my explanation above.
> 
> I see Folly uses the dedicated size as a template parameter.

As does LLVM.

> 
> Is it possible to move the dedicated size to a constructor parameter...so the size only appears on the line that creates the vector...instead of everywhere the vector is passed?

In LLVM, this problem is solved by having `SmallVector` derive from a base, without the inline size NTTP, that has the guts of the vector implementation. Then references to the base class are passed around, and only the final `SmallVector` template itself needs the inline size as an NTTP. Strange that Folly doesn't do this too.

Reply:
@cdavis5e Thank you for the idea with the base class. I will investigate this in a few days. It's exactly what's missing to make this a cleaner interface.

@billhollings Yes, MVKVector meets the requirement of contiguous memory after a spill-over. Again all it is, is a simplified version of std::vector.

Anyway, give me a few more days after Christmas and I will try to implement c5davis5e recommendations to avoid using those special class names.

Reply:
Thanks for continuing to look at this. Looking forward to the updates!

Reply:
@aerofly Just wondering if are still planning to rework this.

If you're stuck for time...perhaps we could see if someone else can look at it to finish off this initial fixed size implementation.

It would be nice to be able to start to make use of the fixed allocations you've made possible!

Reply:
@billhollings I apologize for the delay. I have pretty much finished making the suggested changes. I will try to modify my pull request tomorrow. So in general usage is now as follows:

1) `MVKVector<uint32_t> dynamicOffsets`
  Same as using `std::vector<uint32_t> dynamicOffsets`

2) `MVKVector<uint32_t, 4> dynamicOffsets`
  Reserves 4 elements of uint32_t within the class itself.
  If size of vector stays <= 4 no heap allocations take place.
  If dynamicOffsets grows to > 4 elements, all elements are taken from heap.
  All elements are always in contiguous memory.
  
3) When passing around a MVKVector to a function, one declares the function as follows
  `void function( MVKVectorBase<uint32_t> &vector)`
  `{`
  `  // do something with vector...`
  `  vector.push_back( 13 );`
  `}`
  `  `
  `MVKVector<uint32_t, 4> dynamicOffsets;`
  `function( dynamicOffsets );`

So the only quirk is the fact that you have to use the name 'MVKVectorBase' when passing around MVKVector's to functions. This avoids your previous comment of having to know the number of pre allocated elements when calling a function, but you have to use this special name now.

I hope that's ok.

Reply:
Great. This is looking a bit more generic now. Thanks.

My only suggestion at this point is to wonder if we could flip the names? Make `MVKVector` the name of the base class...and something like `MVKVectorInline` the name of the pre-allocation subclass? 

That would result in:

```
void function( MVKVector<uint32_t> &vector)
{
// do something with vector...
vector.push_back( 13 );
}

MVKVectorInline<uint32_t, 4> dynamicOffsets;
function( dynamicOffsets );
```

That makes the generic  name `MVKVector` the thing that appears everywhere as it passed around to functions...and the specialized name `MVKVectorInline` only appears once...and only where the inline-allocation parameter that defines it appears.

Reply:
Sounds good. Let me see if I can implement this.

Reply:
Ok, I uploaded the new version of MVKVector alongside some code changes. Please review my changes.

You might better know what values to use for MVKVectorInline in certain places.

Reply:
I, uh, think you need to merge with `HEAD` again.

This seems to have previous changes that were merged to `HEAD` between last time and now included in it. Maybe you need to rebase?

Reply:
@aerofly 

Yeah...there is a fair bit of black art to git.

To keep your repo in sync with this repo...first...if you haven't already done it...add the Khronos repo as a remote to your local repo:

    git remote add upstream https://github.com/KhronosGroup/MoltenVK.git

You only have to do this once...and you can verify it via `git remote -v`...which will list all your remotes. This list will now include `origin` (yours on Github) and `upstream` (Khronos on Github).

Then...once you have committed to your local repo...you can periodically do:

    git pull upstream master

...to merge the current state of `upstream` into your local repo.

Depending on the overall order of commits...this will likely trigger a sync commit...which is fine. And you may or may not have to resolve conflicts if needed.

Then push to your GitHub repo as normal via `git push`.

Reply:
@billhollings Thank you very much, I am slowly understanding how its supposed to work. Quite a bit different workflow than I am used to.
So my pull request should be ok now, at least it seems to work on my side.
If you want me to perform any other changes, please let me know.